cpp/src/strings/split/split.cu

```diff
@@ -110,135 +81,442 @@ struct token_counter_fn
+        d_counts[str_idx-1] = ((max_tokens > 0) && (delim_count+1 > max_tokens))
+                              ? max_tokens : delim_count+1;
     }
+
+    rsplit_tokenizer_fn( column_device_view const& d_strings, string_view const& d_delimiter, size_type max_tokens )
+    : base_split_tokenizer(d_strings,d_delimiter,max_tokens) {}
 };
 
+/**
+ * @brief Generic split function called by split() and rsplit().
+ *
+ * This function will first count the number of delimiters in the entire strings
+ * column. Next it records the position of all the delimiters. These positions
+ * are used for the remainder of the code to build string_index_pair elements
+ * for each output column.
+ *
+ * The number of tokens for each string is computed by analyzing the delimiter
+ * position values and mapping them to each string.
+ * The number of output columns is determined by the string with the most tokens.
+ * Next the string_index_pairs for the entire column is created using the
+ * delimiter positions and their string indices vector.
+ *
+ * Finally, each column is built by creating a vector of tokens (string_index_pairs)
+ * according to their position in each string. The first token from each string goes
+ * into the first output column, the 2nd token from each string goes into the 2nd
+ * output column, etc.
+ *
+ * @tparam Tokenizer provides unique functions for split/rsplit.
+ * @param strings_column The strings to split
+ * @param tokenizer Tokenizer for counting and producing tokens
+ * @return table of columns for the output of the split
+ */
+template<typename Tokenizer>
+std::unique_ptr<experimental::table> split_fn( strings_column_view const& strings_column,
+                                               Tokenizer tokenizer,
+                                               rmm::mr::device_memory_resource* mr,
+                                               cudaStream_t stream )
+{
+    std::vector<std::unique_ptr<column>> results;
+    auto strings_count = strings_column.size();
+    if( strings_count == 0 )
+    {
+        results.push_back(make_empty_strings_column(mr,stream));
+        return std::make_unique<experimental::table>(std::move(results));
+    }
+
+    auto execpol = rmm::exec_policy(stream);
+    auto d_offsets = strings_column.offsets().data<int32_t>();
+    d_offsets += strings_column.offset(); // nvbug-2808421 : do not combine with the previous line
+    auto chars_bytes = thrust::device_pointer_cast(d_offsets)[strings_count]
+                     - thrust::device_pointer_cast(d_offsets)[0];
+
+    // count the number of delimiters in the entire column
+    size_type delimiter_count = thrust::count_if( execpol->on(stream),
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(chars_bytes),
+        [tokenizer, d_offsets, chars_bytes] __device__ (size_type idx) {
+            return tokenizer.is_delimiter(idx,d_offsets,chars_bytes);
```

Comment:
Good catch. I thought I had one since I was debugging this for awhile. Anyway, added a couple. Here is one: https://github.com/rapidsai/cudf/blob/520e29c8621061b317712c040312173675ce0746/cpp/tests/strings/split_tests.cpp#L67-L68